exec/java-exec/src/main/java/org/apache/drill/exec/planner/fragment/SimpleParallelizerMultiPlans.java

```diff
@@ -0,0 +1,222 @@
+
+    List<QueryWorkUnit> workUnits = Lists.newArrayList();
+    int plansCount = 0;
+    DrillbitEndpoint[] endPoints = null;
+    long initialAllocation = 0;
+    long maxAllocation = 0;
+
+    final Iterator<Wrapper> iter = planningSet.iterator();
+    while (iter.hasNext()) {
+      Wrapper wrapper = iter.next();
+      Fragment node = wrapper.getNode();
+      boolean isLeafFragment = node.getReceivingExchangePairs().size() == 0;
+      final PhysicalOperator physicalOperatorRoot = node.getRoot();
+      // get all the needed info from leaf fragment
+      if ( (physicalOperatorRoot instanceof Exchange) &&  isLeafFragment) {
+        // need to get info about
+        // number of minor fragments
+        // assignedEndPoints
+        // allocation
+        plansCount = wrapper.getWidth();
+        initialAllocation = (wrapper.getInitialAllocation() != 0 ) ? wrapper.getInitialAllocation()/plansCount : 0;
+        maxAllocation = (wrapper.getMaxAllocation() != 0 ) ? wrapper.getMaxAllocation()/plansCount : 0;
+        endPoints = new DrillbitEndpoint[plansCount];
+        for (int mfId = 0; mfId < plansCount; mfId++) {
+          endPoints[mfId] = wrapper.getAssignedEndpoint(mfId);
+        }
+      }
+    }
+    if ( plansCount == 0 ) {
+      // no exchange, return list of single QueryWorkUnit
+      workUnits.add(generateWorkUnit(options, foremanNode, queryId, reader, rootNode, planningSet, session, queryContextInfo));
+      return workUnits;
+    }
+
+    for (Wrapper wrapper : planningSet) {
+      Fragment node = wrapper.getNode();
+      final PhysicalOperator physicalOperatorRoot = node.getRoot();
+      if ( physicalOperatorRoot instanceof Exchange ) {
+        // get to 0 MajorFragment
+        continue;
+      }
+      boolean isRootNode = rootNode == node;
+
+      if (isRootNode && wrapper.getWidth() != 1) {
+        throw new ForemanSetupException(String.format("Failure while trying to setup fragment. " +
+                "The root fragment must always have parallelization one. In the current case, the width was set to %d.",
+                wrapper.getWidth()));
+      }
+      // this fragment is always leaf, as we are removing all the exchanges
+      boolean isLeafFragment = true;
+
+      // Create a minorFragment for each major fragment.
+      for (int minorFragmentId = 0; minorFragmentId < plansCount; minorFragmentId++) {
+        // those fragments should be empty
+        List<PlanFragment> fragments = Lists.newArrayList();
+
+        PlanFragment rootFragment = null;
+        FragmentRoot rootOperator = null;
+
+        IndexedFragmentNode iNode = new IndexedFragmentNode(minorFragmentId, wrapper);
+        wrapper.resetAllocation();
+        // two visitors here
+        // 1. To remove exchange
+        // 2. To reset operator IDs as exchanges were removed
+        PhysicalOperator op = physicalOperatorRoot.accept(ExchangeManipulatorMaterializerVisitor.INSTANCE, iNode).
```

Comment:
Why ExchangeRemoverVisitor would not drop any exchange? It does drop in any case. Decision on whether to drop or not is done before that visitor is called